harness/determined/cli/trial.py

```diff
@@ -181,7 +181,15 @@ def download(args: Namespace) -> None:
             raise ValueError(f"No checkpoints found for trial {args.trial_id}")
         checkpoint = checkpoints[0]
 
-    path = checkpoint.download(path=args.output_dir)
+    try:
+        path = checkpoint.download(path=args.output_dir)
+    except errors.CheckpointStateException:
+        print(
+            f"Invalid Checkpoint State for checkpoint {checkpoint.uuid}. "
+            f"This checkpoint is in state {checkpoint.state}. "
+            "Please retry with a different checkpoint UUID."
+        )
+        return
```

Comment:
Might also be worth looking at [how the CLI errors out](https://github.com/determined-ai/determined/blob/e9a457d75a021fd326bf2d970a0fe4f8a9062070/harness/determined/cli/cli.py#L291) in other situations and adopting that pattern. I dunno -- it's been a while since I've looked at it. Just a suggestion.

Reply:
I like that, updated to reflect that style.

Reply:
Looking into `determined.cli.cli` here, it'd be nice if the raised exception could be localized there to exactly this problem. But I'm not sure it makes sense to try and hack that in. Probably what you decided, too.

I _do_ think there's some value in using `cli.die`. Value I see includes:
* Standardization -- when exiting via `die` we get pretty colors
* explicit "something when wrong"
* nonzero exit code

But it's a little weird that most of the dying is done in `cli.cli` and this would be the first(?) example of it elsewhere.

Have you thought about using `cli.die` here, yourself? What did you conclude?

Reply:
Yeah, let's keep specific exceptions/error messages local to their respective CLI methods.

I also concluded that doing `cli.die` outside of `cli.cli` feels weird, but that definitely can change. I'd rather do standardizations like that in another larger and wide-reaching PR, especially given that this current PR is not about long-term strategy but rather a fix for a particular bug.

Reply:
There's some weirdness for true. `cli.cli` thinks that it generally has all the necessary context to surface errors, but `cli.trial.download` thinks that it has context `cli.cli` doesn't. It's too bad if we can't satisfy `cli.cli` there in keeping with the existing standard, but I'm not going to lose too much sleep over it. I dunno. It _does_ feel like `cli.cli` is the right place to die rather than here.

No matter, I don't think keeping exceptions/error messages local to their respective CLI methods is the current standard.

If you don't want to import from `cli.cli` and you don't want to propagate this error to `cli.cli`, how do you feel about factoring `die` out of `cli.cli` and into `cli._util`? I think it's a slightly weird experience for users to see some errors printed in red and resulting in nonzero exit codes and other errors printed

Reply:
As communicated OOB, I think it's important that an error here is handled the same way as other CLI errors are at the bottom of `cli.cli`. `download` _does_ have context here that `cli.cli` doesn't, but that context can be passed to `cli.cli` by catching and reraising the error.